### PR TITLE
chore: centralize swapper into sdk-like set of functions

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Preparation for #5024 .

Centralises all calls to swappers, simplifying where possible. Trade execution is now contained within a single event emitter, which combines 3 swapper function calls and orchestration into a single component. 

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

Preparation for #5024, but does not close it.

## Risk

High risk of broken logic and orchestration for trade input, trade quotes, trade execution for all swappers. Numbers and calculations should be fine, but the general way the app behaves in the context of trades could be broken.

## Testing

Please do a complete test of all swappers, focusing on behaviour and logic - numbers should be fine but give a sanity check just in case.

### Engineering

See above. 

### Operations

See above.

## Screenshots (if applicable)

Thor trade
image
image

0x trade
image
image

osmosis trade
image
image

1inch trade
image
image

CoW
image
image

LiFi
image
image

approvals still working
image
image

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5035** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5035" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5035?utm_source=stack-comment).

File: src/components/MultiHopTrade/hooks/useSupportedAssets.tsx [hunk 1]
```diff
@@ -1,60 +1,46 @@
...
 import { useAppSelector } from 'state/store'
 
 export const useSupportedAssets = () => {
   const sellAsset = useAppSelector(selectSellAsset)
-  const nonNftAssetIds = useAppSelector(selectNonNftAssetIds)
   const sortedAssets = useAppSelector(selectAssetsSortedByMarketCapUserCurrencyBalanceAndName)
-  const { LifiSwap, ThorSwap, ZrxSwap, OneInch, Cowswap, OsmosisSwap } =
-    useAppSelector(selectFeatureFlags)
+  const featureFlags = useAppSelector(selectFeatureFlags)
   useAppSelector(selectFeatureFlags)
```
Comment (kaladinlight) [inline src/components/MultiHopTrade/hooks/useSupportedAssets.tsx:14]: cleanup while we in here
```suggestion
```

File: src/components/MultiHopTrade/hooks/useSupportedAssets.tsx [hunk 2]
```diff
@@ -1,60 +1,46 @@
...
-    if (ZrxSwap) result.push(zrxSwapper)
-    if (OneInch) result.push(oneInchSwapper)
-    if (Cowswap) result.push(cowSwapper)
-    if (OsmosisSwap) result.push(osmosisSwapper)
-    return result
-  }, [Cowswap, LifiSwap, OneInch, ThorSwap, ZrxSwap, OsmosisSwap])
+  const enabledSwappers = useMemo(() => getEnabledSwappers(featureFlags, false), [featureFlags])
 
   const [supportedSellAssets, setSupportedSellAssets] = useState<Asset[]>([])
+  const [supportedSellAssetsIds, setSupportedSellAssetsIds] = useState<Set<AssetId>>(new Set())
```
Comment (kaladinlight) [inline src/components/MultiHopTrade/hooks/useSupportedAssets.tsx:19]:
```suggestion
  const [supportedSellAssetIds, setSupportedSellAssetIds] = useState<Set<AssetId>>(new Set())
```

File: src/components/MultiHopTrade/hooks/useSupportedAssets.tsx [hunk 3]
```diff
@@ -1,60 +1,46 @@
...
   const [supportedBuyAssets, setSupportedBuyAssets] = useState<Asset[]>([])
+  const [supportedBuyAssetsIds, setSupportedBuyAssetsIds] = useState<Set<AssetId>>(new Set())
```
Comment (kaladinlight) [inline src/components/MultiHopTrade/hooks/useSupportedAssets.tsx:21]:
```suggestion
  const [supportedBuyAssetIds, setSupportedBuyAssetIds] = useState<Set<AssetId>>(new Set())
```

File: src/lib/swapper/tradeExecution.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import { TxStatus } from '@shapeshiftoss/unchained-client'
+import EventEmitter from 'events'
+import { TRADE_POLL_INTERVAL_MILLISECONDS } from 'components/MultiHopTrade/hooks/constants'
+import { withFromOrXpub } from 'components/MultiHopTrade/hooks/useTradeExecution/helpers'
+import { poll } from 'lib/poll/poll'
+
+import { swappers } from './constants'
+import type { TradeExecutionInput } from './types'
+
+export class TradeExecution extends EventEmitter {
```
Comment (kaladinlight) [inline src/lib/swapper/tradeExecution.ts:10]: would be nice to have typed events/listeners for client consumption as a stretch, and I think returning an object might be easier to make sense that the cancel function is being returned (don't get much from the return type to go off).

File: src/lib/swapper/swapper.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        console.error(result.reason)
+      }
+      return isFulfilled
+    })
+    .map(result => (result as PromiseFulfilledResult<QuoteResult>).value)
+
+  return successfulQuotes
+}
+
+export const getSupportedSellAssets = async (enabledSwappers: SwapperName[]) => {
```
Comment (0xApotheosis) [inline src/lib/swapper/swapper.ts:47]:
```suggestion
export const getSupportedSellAssetIds = async (enabledSwappers: SwapperName[]) => {
```

File: src/lib/swapper/swapper.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  const { assets } = new AssetService()
+  const supportedAssetIds = await Promise.all(
+    swappers
+      .filter(({ swapperName }) => enabledSwappers.includes(swapperName))
+      .map(({ swapper }) => swapper.filterAssetIdsBySellable(assets)),
+  )
+  return new Set(supportedAssetIds.flat())
+}
+
+export const getSupportedBuyAssets = async (enabledSwappers: SwapperName[], sellAsset: Asset) => {
```
Comment (0xApotheosis) [inline src/lib/swapper/swapper.ts:57]:
```suggestion
export const getSupportedBuyAssetIds = async (enabledSwappers: SwapperName[], sellAsset: Asset) => {
```

File: src/lib/swapper/swapper.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import { AssetService } from 'lib/asset-service'
+import { isFulfilled as isFulfilledPredicate, timeout } from 'lib/utils'
+
+import type { GetTradeQuoteInput, SwapErrorRight, SwapperName, TradeQuote2 } from './api'
+import { QUOTE_TIMEOUT_ERROR, QUOTE_TIMEOUT_MS, swappers } from './constants'
+import type { QuoteResult, TradeQuoteDeps } from './types'
+
+export { TradeExecution } from './tradeExecution'
+
+// gets trade quotes
```
Comment (0xApotheosis) [inline src/lib/swapper/swapper.ts:11]: Function name probably speaks for itself.

```suggestion
```

File: src/lib/swapper/api.ts [hunk 8]
```diff
@@ -149,6 +151,19 @@ export enum SwapperName {
   OneInch = '1INCH',
 }
 
+export const ALL_SWAPPERS = [
```
Comment (0xApotheosis) [inline src/lib/swapper/api.ts:154]: We don't seem to use this guy anywhere.

In any case it should probably be programmatic so we don't forget to update it:

```ts
export const ALL_SWAPPERS = Object.values(SwapperName).filter((value): value is SwapperName => 
  Object.values(SwapperName).includes(value) && value !== SwapperName.Test
);
```

File: src/lib/swapper/api.ts [hunk 9]
```diff
@@ -149,6 +151,19 @@ export enum SwapperName {
...
+  SwapperName.CowSwap,
+  SwapperName.LIFI,
+  SwapperName.OneInch,
+  SwapperName.Osmosis,
+  SwapperName.Thorchain,
+  SwapperName.Zrx,
+]
+
+export type SwapperConfig = {
```
Comment (0xApotheosis) [inline src/lib/swapper/api.ts:163]: Likewise, not used - though I assume there are future plans for it?

File: src/lib/swapper/tradeExecution.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            quoteBuyAssetAccountId,
+            getState,
+          })
+
+          const payload = { stepIndex, status, message, buyTxHash }
+
+          this.emit('status', payload)
+
+          if (status === TxStatus.Confirmed) this.emit('success', payload)
+          if (status === TxStatus.Failed) this.emit('fail', payload)
```
Comment (0xApotheosis) [inline src/lib/swapper/tradeExecution.ts:79]: Should we enum these trade execution event strings?

File: src/state/slices/assetsSlice/assetsSlice.ts [hunk 11]
```diff
@@ -88,7 +88,7 @@ export const makeAsset = (minimalAsset: MinimalAsset): Asset => {
   const explorerLinks = ((): ExplorerLinks => {
     const feeAssetId = chainIdToFeeAssetId(chainId)
     if (!feeAssetId) throw new Error('makeAsset: feeAssetId not found')
-    const feeAsset = getAssetService().getAll()[feeAssetId]
+    const feeAsset = getAssetService().assetsById[feeAssetId]
```
Comment (0xApotheosis) [inline src/state/slices/assetsSlice/assetsSlice.ts:91]: 🧠

**Review by 0xApotheosis — COMMENTED** [review]
First pass, nothing blocking - looking very sane.

Will have another architectural pass now.

File: src/state/apis/swappers/swappersApi.ts [hunk 12]
```diff
@@ -1,43 +1,40 @@
...
 import { selectFeatureFlags } from 'state/slices/selectors'
 
 import { BASE_RTK_CREATE_API_CONFIG } from '../const'
-import { getCowTradeQuoteHelper } from './helpers/getCowTradeQuoteHelper'
-import { getOneInchTradeQuoteHelper } from './helpers/getOneInchTradeQuoteHelper'
-import { getOsmosisTradeQuoteHelper } from './helpers/getOsmosisTradeQuoteApiHelper'
-import { getZrxTradeQuoteHelper } from './helpers/getZrxTradeQuoteHelper'
 
-const QUOTE_TIMEOUT_MS = 10_000
+const getDependencies = (state: ReduxState, getTradeQuoteInput: GetTradeQuoteInput) => {
```
Comment (0xApotheosis) [inline src/state/apis/swappers/swappersApi.ts:17]: Type my return 🙏

**Review by 0xApotheosis — APPROVED** [review]
Conceptual stamp, pending ops functional sign-off.

Feel free to address the non-blocking comments as you feel appropriate.

File: src/lib/swapper/tradeExecution.ts (hunk 4 above)
Comment (woodenfurniture) [reply src/lib/swapper/tradeExecution.ts:10]: awesome suggestions, will do

File: src/lib/swapper/tradeExecution.ts (hunk 10 above)
Comment (woodenfurniture) [reply src/lib/swapper/tradeExecution.ts:79]: will do in follow-up PR

**Pull request merged**